### PR TITLE
Implement sensitivity analysis server

### PR DESCRIPTION
**Pull request by FranckLecuyer**

### CODE REVIEW

**Review by niconoir — COMMENTED** [review]
Minor remarks.

File: src/main/java/org/gridsuite/sensitivityanalysis/server/RestTemplateConfig.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import com.powsybl.sensitivity.json.SensitivityJsonModule;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.http.converter.HttpMessageConverter;
+import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder;
+import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
+import org.springframework.web.client.RestTemplate;
+
+@Configuration
+public class RestTemplateConfig {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/sensitivityanalysis/server/RestTemplateConfig.java:30]: Done

File: src/main/java/org/gridsuite/sensitivityanalysis/server/entities/GlobalStatusEntity.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright (c) 2020, RTE (http://www.rte-france.com)
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/sensitivityanalysis/server/entities/GlobalStatusEntity.java]: Done

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/NotificationService.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/sensitivityanalysis/server/service/NotificationService.java]: Done

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInput.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    private void buildSensitivityVariableSets() {
+        List<List<IdentifiableAttributes>> variablesFiltersLists = context.getVariablesFiltersListUuids().stream()
+            .map(variablesFilterUuid -> filterService.getIdentifiablesFromFilter(variablesFilterUuid, context.getNetworkUuid(), context.getVariantId()))
+            .collect(Collectors.toList());
+
+        variablesFiltersLists.forEach(variablesList -> {
+            List<WeightedSensitivityVariable> variables = new ArrayList<>();
+            if (!variablesList.isEmpty()) {
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInput.java:62]: The sum is not needed, a weight is expected.This is not necessary to nomalize.

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisWorkerService.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+    private ObjectMapper objectMapper;
+
+    private Map<UUID, CompletableFuture<SensitivityAnalysisResult>> futures = new ConcurrentHashMap<>();
+
+    private Map<UUID, SensitivityAnalysisCancelContext> cancelComputationRequests = new ConcurrentHashMap<>();
+
+    private Set<UUID> runRequests = Sets.newConcurrentHashSet();
+
+    private Lock lockRunAndCancelSensitivityAnalysis = new ReentrantLock();
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisWorkerService.java:79]: final?

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/sensitivityanalysis/server/SensitivityAnalysisController.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                                                         @Parameter(description = "Quad filters list uuids") @RequestParam(name = "quadFiltersListUuid", required = false) List<UUID> quadFiltersListUuids,
+                                                         @Parameter(description = "Provider") @RequestParam(name = "provider", required = false) String provider,
+                                                         @Parameter(description = "reportUuid") @RequestParam(name = "reportUuid", required = false) UUID reportUuid,
+                                                         @RequestBody(required = false) SensitivityAnalysisParameters parameters) {
+        SensitivityAnalysisParameters nonNullParameters = getNonNullParameters(parameters);
+        List<UUID> nonNullOtherNetworkUuids = getNonNullOtherNetworkUuids(otherNetworkUuids);
+        SensitivityAnalysisResult result = workerService.run(new SensitivityAnalysisRunContext(networkUuid, variantId, nonNullOtherNetworkUuids, variablesFiltersListUuids, contingencyListUuids, quadFiltersListUuids, null, provider, nonNullParameters, reportUuid));
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/sensitivityanalysis/server/SensitivityAnalysisController.java]: Done

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInput.java (hunk 4 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisInput.java:62]: Done

File: src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisWorkerService.java (hunk 5 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/sensitivityanalysis/server/service/SensitivityAnalysisWorkerService.java]: Done

**Review by geofjamg — APPROVED** [review]

**Pull request merged**